pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+    >>> tape.jacobian(dev)
+    [[-0.45478169]]
+    """
+
+    _cast = staticmethod(np.array)
+
+    def __init__(self, name=None):
+        super().__init__()
+        self.name = name
+
+        self._prep = []
+        """list[~.Operation]: Tape state preparations."""
+
+        self._ops = []
+        """list[~.Operation]: quantum operations recorded by the tape."""
+
+        self._obs = []
+        """list[tuple[~.MeasurementProcess, ~.Observable]]: measurement processes and
+        coresponding observables recorded by the tape."""
+
+        self._par_info = {}
+        """dict[int, dict[str, Operation or int]]: Parameter information. Keys are
+        parameter indices (in the order they appear on the tape), and values are a
+        dictionary containing the corresponding operation and operation parameter index."""
+
+        self._trainable_params = set()
+        self._graph = None
+        self._output_dim = 0
+
+        self.wires = qml.wires.Wires([])
+        self.num_wires = 0
+
+        self.jacobian_options = {}
+
+        self.hash = 0
+        self.is_sampled = False
+
+        self._stack = None
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: wires={self.wires}, params={self.num_params}>"
+
+    def __enter__(self):
+        if not QueuingContext.recording():
+            # if the tape is the first active queuing context
+            # monkeypatch the operations to support the new queuing context
+            with contextlib.ExitStack() as stack:
+                for mock in mock_operations():
+                    stack.enter_context(mock)
+                self._stack = stack.pop_all()
+
+        QueuingContext.append(self)
+        return super().__enter__()
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        super().__exit__(exception_type, exception_value, traceback)
+
+        if not QueuingContext.recording():
+            # remove the monkeypatching
+            self._stack.__exit__(exception_type, exception_value, traceback)
+
+        self._process_queue()
+
+    @property
+    def interface(self):
+        """str, None: automatic differentiation interface used by the quantum tap (if any)"""
+        return None
+
+    # ========================================================
+    # construction methods
+    # ========================================================
+
+    def _process_queue(self):
+        """Process the annotated queue, creating a list of quantum
+        operations and measurement processes.
+
+        This method sets the following attributes:
+
+        * ``_ops``
+        * ``_obs``
+        * ``_par_info``
+        * ``_output_dim``
+        * ``_trainable_params``
+        * ``is_sampled``
+        """
+        self._prep = []
+        self._ops = []
+        self._obs = []
+        self._output_dim = 0
+
+        for obj, info in self._queue.items():
+
+            if isinstance(obj, QuantumTape):
+                self._ops.append(obj)
```

Comment:
Nothing will happen until later when `expand` is called 👍 

I went back and forth on this a lot, and I think this is the approach that makes the most sense. Nesting only works if the expansion is delayed, since _how_ we do the expansion depends on external variables (the device we are using, the gradient method, etc).

In the current approach, later when `tape.expand()` is called, this tape will be expanded, validation will run, and an error will be raised ('Cannot have measurements in a non-terminal position' etc.).